ocaml/utils/global.ml

```diff
@@ -0,0 +1,146 @@
+      Format.fprintf ppf "@[<hov 1>%s%a@]"
+        head
+        (pp_concat print_arg_pair) args
+    and print_arg_pair ppf (name, arg) =
+      Format.fprintf ppf "[%a:%a]" print name print arg
+
+    let output = print |> Misc.output_of_print
+
+    let rec hash { head; args } =
+      Hashtbl.hash (head, List.map hash_arg args)
+    and hash_arg (name, param) =
+      Hashtbl.hash (hash name, hash param)
+  end)
+end
+
+type t = {
+  head : string;
+  args : (Name.t * t) list;
+  params : (Name.t * t) list;
+}
+
+include Identifiable.Make (struct
+  type nonrec t = t
+
+  let rec compare
+      { head = head1; args = args1; params = params1 }
+      { head = head2; args = args2; params = params2 } =
+    match String.compare head1 head2 with
+    | 0 -> begin
+        match List.compare compare_pairs args1 args2 with
+        | 0 -> List.compare compare_pairs params1 params2
+        | c -> c
+      end
+    | c -> c
+  and compare_pairs (param1, value1) (param2, value2) =
+    match Name.compare param1 param2 with
+    | 0 -> compare value1 value2
+    | c -> c
+
+  let equal t1 t2 = t1 == t2 || compare t1 t2 = 0
+
+  let rec print ppf { head; args; params } =
+    Format.fprintf ppf "@[<hov 1>%s%a%a@]"
+      head
+      (pp_concat print_arg_pair) args
+      (pp_concat print_param_pair) params
+  and print_arg_pair ppf (name, value) =
+    Format.fprintf ppf "[%a:%a]" Name.print name print value
+  and print_param_pair ppf (name, value) =
+    Format.fprintf ppf "{%a:%a}" Name.print name print value
+
+  let output = print |> Misc.output_of_print
+
+  let rec hash { head; args; params } =
+    Hashtbl.hash (head, List.map hash_pair args, List.map hash_pair params)
+  and hash_pair (param, value) =
+    Hashtbl.hash (Name.hash param, hash value)
+end)
+
+module Subst = Name.Map
+type subst = t Subst.t
+
+(* CR-someday lmaurer: Should try and make this unnecessary or at least cheap.
+   Could do it by making [Name.t] an unboxed existential so that converting from
+   [t] is the identity. Or just have [Name.t] wrap [t] and ignore [params]. *)
+let rec to_name ({ head; args; params = _ }) : Name.t =
+  { head; args = List.map arg_to_name args }
+and arg_to_name (name, value) =
+  name, to_name value
+
+let union_args args1 args2 =
+  (* [Misc.Stdlib.List.merge_map] of empty will iterate through the whole
+      other list, so special-case that out here *)
+  match args1, args2 with
+  | [], args | args, [] -> args
+  | _, _ ->
+      Misc.Stdlib.List.merge_map args1 args2
+        ~cmp:(fun (name1, _) (name2, _) -> Name.compare name1 name2)
+        ~left_only:(fun pair1 -> pair1)
+        ~right_only:(fun pair2 -> pair2)
+        ~both:
+          (fun (param, value1) (_param, value2) ->
+            Misc.fatal_errorf
+              "Duplicate arguments for %a:@ %a@ vs. %a"
+              Name.print param
+              print value1
+              print value2)
+
+let rec subst ({ head; args; params } as param) (s : subst) =
```

Comment:
I added the code to determine whether it's changed, then realized it's just as easy to return the original argument if it hasn't. Still worth returning a separate bool (in a field with a descriptive name) rather than letting the caller use `phys_equal`?

Reply:
Oh heh, not sure how this ended up as two conversations, but yes, `subst` now returns ``[ `Changed | `Did_not_change ]`` in addition to the result.